package_scanner.nim

```diff
@@ -1,135 +1,367 @@
+## - Try to Git clone the repo (deletes folder after).
+## - Works online/offline
+##
+## Use
+## ---
+##
+## - Off-line: ``nim c -r package_scanner2.nim`` (Faster, wont use Internet).
+## - On-line: ``nim c -d:ssl -r package_scanner2.nim`` (Checks URLs).
+##
+## Credits
+## -------
+##
+## Based and inspired on a previous ``package_scanner.nim`` from:
+## Copyright 2015 Federico Ceratto <federico.ceratto@gmail.com>
+## Released under GPLv3 License, see /usr/share/common-licenses/GPL-3
+
+import
+  httpclient, json, net, os, osproc, sets, strutils, streams, logging,
+  unittest, sequtils, std/editdistance
 
-# A very simple Nim package scanner.
-#
-# Scans the package list from this repository.
-#
-# Check the packages for:
-#  * Missing name
-#  * Missing/unknown method
-#  * Missing/unreachable repository
-#  * Missing tags
-#  * Missing description
-#  * Missing/unknown license
-#  * Insecure git:// url on GitHub
-#
-# Usage: nim c -d:ssl -r package_scanner.nim
-#
-# Copyright 2015 Federico Ceratto <federico.ceratto@gmail.com>
-# Released under GPLv3 License, see /usr/share/common-licenses/GPL-3
-
-import httpclient
-import net
-import json
-import os
-import sets
-import strutils
 
 const
-  LICENSES = @[
-    "Allegro 4 Giftware",
-    "Apache License 2.0",
-    "BSD",
-    "BSD2",
-    "BSD3",
-    "CC0",
-    "GPL",
-    "GPLv2",
-    "GPLv3",
-    "LGPLv2",
-    "LGPLv3",
-    "MIT",
-    "MS-PL",
-    "MPL",
-    "WTFPL",
+  allowBrokenUrl = true      ## Allow broken Repo URLs
+  allowMissingNimble = true  ## Allow missing ``*.nimble`` files
+  checkNimbleFile = true     ## Check if repos have ``*.nimble`` file
+  checkProfanity = false     ## Check for Profanities
+  checkByGit = false         ## Check via Git
+  httpTimeout = 10_000       ## Timeout. Below ~2000 false positives happen?
+  tagsMaxLen = 32            ## Maximum lenght for the string in tags
+  tagsMaximum = 17           ## Maximum number of tags allowed on the list
+  lineLenMax = 250           ## Maximum line lenght for values
+  profanityLevensteinMax = 1 ## Levenstein Distance algo threshold
+  vcsTypes = ["git", "hg"]   ## Valid known Version Control Systems.
+  tagsBlacklist = ["nimrod", "nim"] ## Tags that should be not allowed.
+  keysRequired = ["name", "url", "method", "tags", "description", "license"]
+  gitTempDir = "/tmp/gitTempDir"
+  gitCmd = "git clone --no-checkout --no-tags --bare --depth 1 $1 " & gitTempDir
+  nodeCmd = "node ./validate_json.js"
+  pythonCmd = "python3 ./validate_json.py"
+  rubyCmd = "ruby ./validate_json.rb"
+  defaultFilePermissions = {fpUserWrite, fpUserRead, fpGroupWrite, fpGroupRead, fpOthersRead}
+  packagesFilePath = currentSourcePath().parentDir / "packages.json"
+  packagesJsonStr = readFile(packagesFilePath)
+  hostsSkip = [
+    "https://bitbucket",
+    "https://gitlab.3dicc",
+    "https://mahlon@bitbucket",
+    "https://notabug",
+  ] ## Hostnames to skip checks, for reliability, wont like direct HTTP GET?.
```

Comment:
These are strange URLs, they don't contain a real hostname. I assume you are parsing them in some custom way? Why not make this more clear with:

```nim
hostsSkip = [
  "bitbucket.org",
  "gitlab.3dicc",
  ...
]
```